### PR TITLE
Use a constant empty tar header to avoid extra allocations

### PR DESCRIPTION
**Pull request by segiddins**

### CODE REVIEW

**Review by martinemde — APPROVED** [review]
Nice.

File: lib/rubygems/package/tar_header.rb [hunk 1]
```diff
@@ -172,6 +172,15 @@ def initialize(vals)
     @empty = vals[:empty]
   end
 
+  EMPTY = new({ # :nodoc:
+    mode: 0,
+    name: "",
+    prefix: "",
+    size: 0,
+    empty: true,
+  }).freeze
```
Comment (martinemde) [inline lib/rubygems/package/tar_header.rb:181]: To really be identical you need the following. The default values are not the same as what an empty header outputs.

```suggestion
  EMPTY = new({ # :nodoc:
    checksum: 0,
    gname: "",
    linkname: "",
    magic: "",
    mode: 0,
    name: "",
    prefix: "",
    size: 0,
    uname: "",
    version: 0,
    empty: true,
  }).freeze
```
Here's my irb doublecheck:

```ruby
irb(main):033> EMPTY = Gem::Package::TarHeader.new({ checksum: 0, gname: "", linkname: "", magic: "", mode: 0, name: "", prefix: "", size: 0, uname: "", version: 0, empty: true,  })
=> #<Gem::Package::TarHeader:0x00000001203e60e0 @checksum=0, @devmajor=0, @devminor=0, @empty=true, @gid=0, @gname="", @linkname="", @magic="", @mode=0, @mtime=0, @name="", @prefix="", @size=0, @typeflag="0", @uid=0, @uname="", @version=0>
irb(main):034> Gem::Package::TarHeader.from(StringIO.new(Gem::Package::TarHeader::EMPTY_HEADER))
=> #<Gem::Package::TarHeader:0x00000001204cde90 @checksum=0, @devmajor=0, @devminor=0, @empty=true, @gid=0, @gname="", @linkname="", @magic="", @mode=0, @mtime=0, @name="", @prefix="", @size=0, @typeflag="0", @uid=0, @uname="", @version=0>
```


We should also add a test.

**Review by martinemde — COMMENTED** [review]
I approved it before double checking, here's the additional I think it needs.

**Review by simi — APPROVED** [review]
:thinking: *wondering what's the impact of this change*

**Comment by martinemde** [comment]
@simi are you thinking the tar header is expected to be mutable? That's worth considering...

**Comment by simi** [comment]
> @simi are you thinking the tar header is expected to be mutable? That's worth considering...

No, I was just wondering what's memory footprint impact.

File: lib/rubygems/package/tar_header.rb [hunk 2]
```diff
@@ -241,7 +257,7 @@ def header(checksum = @checksum)
 
     header = header.pack PACK_FORMAT
 
-    header << ("\0" * ((512 - header.size) % 512))
+    header.ljust 512, "\0"
```
Comment (deivid-rodriguez) [inline lib/rubygems/package/tar_header.rb:260]: This seems like a neat change regardless of the rest :)

**Review by deivid-rodriguez — APPROVED** [review]
I also wonder about impact. Do we instantiate an empty tar reader once per gem, and this would reuse it for `gem install` invocations that install many gems?

Anyways, code changes look good to me!

**Comment by segiddins** [comment]
There is one empty tar header per tar file read, in my experience. each gem requires reading 2 tar files.

**Review by deivid-rodriguez — APPROVED** [review]

**Comment by deivid-rodriguez** [comment]
Should we squash into a single commit? Right now there's three commits, two of them with the same commit message, which seems a bit confusing! Other than that, seems great!

**Comment by deivid-rodriguez** [comment]
@segiddins I squashed commits myself and will be merging this, hope it's ok.

**Pull request merged**